crates/ruff_server/src/server/api/requests/execute_command.rs

```diff
@@ -0,0 +1,144 @@
+use serde::Deserialize;
+
+enum Command {
+    Format,
+    FixAll,
+    OrganizeImports,
+}
+
+pub(crate) struct ExecuteCommand;
+
+#[derive(Deserialize)]
+struct TextDocumentArgument {
+    uri: types::Url,
+}
+
+impl super::RequestHandler for ExecuteCommand {
+    type RequestType = req::ExecuteCommand;
+}
+
+impl super::SyncRequestHandler for ExecuteCommand {
+    fn run(
+        session: &mut Session,
+        _notifier: client::Notifier,
+        requester: &mut client::Requester,
+        params: types::ExecuteCommandParams,
+    ) -> server::Result<Option<serde_json::Value>> {
+        let Some(command) = Command::from_str(&params.command) else {
+            return Err(anyhow::anyhow!("")).with_failure_code(ErrorCode::InvalidParams);
+        };
+
+        let mut changes = HashMap::new();
+        let documents =
+            args_as_text_documents(params.arguments).with_failure_code(ErrorCode::InvalidParams)?;
+        for document in documents {
+            let snapshot = session
+                .take_snapshot(&document.uri)
+                .ok_or(anyhow::anyhow!(
+                    "Document snapshot not available for {}",
+                    document.uri
+                ))
+                .with_failure_code(ErrorCode::InternalError)?;
+            match command {
+                Command::FixAll => {
+                    let edits = super::code_action_resolve::fix_all_edit(
+                        snapshot.document(),
+                        &snapshot.configuration().linter,
+                        snapshot.encoding(),
+                    )
+                    .with_failure_code(ErrorCode::InternalError)?;
+                    changes.insert(document.uri, edits);
+                }
+                Command::Format => {
+                    let response = super::format::format_document(&snapshot)?;
+                    if let Some(edits) = response {
+                        changes.insert(document.uri, edits);
+                    }
+                }
+                Command::OrganizeImports => {
+                    let edits = super::code_action_resolve::organize_imports_edit(
+                        snapshot.document(),
+                        &snapshot.configuration().linter,
+                        snapshot.encoding(),
+                    )
+                    .with_failure_code(ErrorCode::InternalError)?;
+                    changes.insert(document.uri, edits);
+                }
+            }
+        }
+
+        if !changes.is_empty() {
+            // check if we can apply a workspace edit
+            if !session.resolved_client_capabilities().apply_edit {
+                return Err(anyhow::anyhow!("Cannot send workspace edit to client: the client does not support `workspace/applyEdit`")).with_failure_code(ErrorCode::InternalError);
+            }
+            apply_edit(
+                requester,
+                command.label(),
+                types::WorkspaceEdit::new(changes),
```

Comment:
Hmm, this could get interesting once we have fixes that can "fix" across file because it is then no longer guaranteed that the changes are not overlapping. But that's something to worry about at another day. 

What we could consider doing is to move this inside the loop?

Reply:
Multi-file fixes are going to require a lot of refactoring regardless, but yes, this is definitely something we'll need to solve eventually. For now though, it's a future concern.